amaranth/sim/pysim.py

```diff
@@ -439,6 +442,25 @@ def _step(self):
             # 2. commit: apply every queued signal change, waking up any waiting processes
             converged = self._state.commit(changed)
 
+    def _step_tb(self):
+        changed = set() if self._vcd_writers else None
+
+        # Run processes waiting for an interval to expire (mainly `add_clock_process()``)
+        self._step_rtl(changed)
+
+        # Run testbenches waiting for an interval to expire, or for a signal to change state
+        converged = False
+        while not converged:
+            converged = True
+            # Schedule testbenches in a deterministic, predictable order by iterating a list
+            for testbench in self._testbenches:
+                if testbench.runnable:
+                    testbench.runnable = False
+                    while testbench.run():
```

Comment:
I believe `process.run()` above should also be wrapped in `while process.run(): pass` so that it doesn't terminate early

Reply:
Early termination only applies to testbenches (it's gated on `if testbench:` in the coroutine process implementation).

Reply:
oops, my bad